File: web/src/types/analytics.ts
/**
 * Types matching `backend/app/schemas/analytics.py`.
 *
 * Wired endpoints on web (slice 4):
 *   - GET /analytics/summary          → AnalyticsSummaryResponse
 *   - GET /analytics/productivity-score → ProductivityScoreResponse
 *   - GET /analytics/time-tracking    → TimeTrackingResponse
 *   - GET /analytics/habit-correlations → HabitCorrelationResponse
 *
 * NOT wired: GET /analytics/project-progress — it requires an integer
 * Postgres project_id, but web projects live in Firestore with string
 * doc IDs. Wiring this properly needs either a backend change or a
 * cross-reference resolver and is out of scope for slice 4.
 */

// ── Productivity score ───────────────────────────────────────────

export interface ScoreBreakdown {
  task_completion: number;
  on_time: number;
  habit_completion: number;
  estimation_accuracy: number;
}

export interface DailyScore {
  /** ISO date `YYYY-MM-DD`. */
  date: string;
  score: number;
  breakdown: ScoreBreakdown;
}

export interface BestWorstDay {
  date: string;
  score: number;
}

export type ProductivityTrend = 'improving' | 'declining' | 'stable';

export interface ProductivityScoreResponse {
  scores: DailyScore[];
  average_score: number;
  trend: ProductivityTrend;
  best_day: BestWorstDay | null;
  worst_day: BestWorstDay | null;
}

// ── Time tracking ─────────────────────────────────────────────────

export type TimeTrackingGroupBy = 'project' | 'tag' | 'priority' | 'day';

export interface TimeTrackingEntry {
  group: string;
  total_minutes: number;
  task_count: number;
  avg_minutes_per_task: number;
  estimated_total: number;
  accuracy_pct: number;
}

export interface TimeTrackingResponse {
  entries: TimeTrackingEntry[];
  total_tracked_minutes: number;
  total_estimated_minutes: number;
  overall_accuracy_pct: number;
  most_time_consuming_project: string | null;
  most_accurate_estimates: string | null;
}

// ── Habit correlations ────────────────────────────────────────────

export type CorrelationDirection = 'positive' | 'negative' | 'neutral';

export interface HabitCorrelation {
  habit: string;
  done_productivity: number;
  not_done_productivity: number;
  correlation: CorrelationDirection;
  interpretation: string;
}

export interface HabitCorrelationResponse {
  correlations: HabitCorrelation[];
  top_insight: string;
  recommendation: string;
}

// ── Summary ──────────────────────────────────────────────────────

export interface TodaySummary {
  completed: number;
  remaining: number;
  score: number;
}

export interface WeekSummary {
  completed: number;
  remaining: number;
  score: number;
  trend: string;
}

export interface MonthSummary {
  completed: number;
  remaining: number;
  score: number;
}

export interface StreakSummary {
  current_productive_days: number;
  longest_productive_days: number;
}

export interface HabitSummaryBucket {
  completion_rate_7d: number;
  completion_rate_30d: number;
}

export interface AnalyticsSummaryResponse {
  today: TodaySummary;
  this_week: WeekSummary;
  this_month: MonthSummary;
  streaks: StreakSummary;
  habits: HabitSummaryBucket;
}

// ── Query param helpers ──────────────────────────────────────────

export interface ProductivityScoreParams {
  period?: 'daily' | 'weekly' | 'monthly';
  start_date?: string;
  end_date?: string;
}

export interface TimeTrackingParams {
  start_date?: string;
  end_date?: string;
  group_by?: TimeTrackingGroupBy;
}

File: backend/app/schemas/analytics.py
from datetime import date
from typing import Optional

from pydantic import BaseModel


# --- Productivity Score ---


class ScoreBreakdown(BaseModel):
    task_completion: float
    on_time: float
    habit_completion: float
    estimation_accuracy: float


class DailyScore(BaseModel):
    date: date
    score: float
    breakdown: ScoreBreakdown


class BestWorstDay(BaseModel):
    date: date
    score: float


class ProductivityScoreResponse(BaseModel):
    scores: list[DailyScore]
    average_score: float
    trend: str  # "improving", "declining", "stable"
    best_day: Optional[BestWorstDay] = None
    worst_day: Optional[BestWorstDay] = None


# --- Time Tracking ---


class TimeTrackingEntry(BaseModel):
    group: str
    total_minutes: int
    task_count: int
    avg_minutes_per_task: float
    estimated_total: int
    accuracy_pct: float


class TimeTrackingResponse(BaseModel):
    entries: list[TimeTrackingEntry]
    total_tracked_minutes: int
    total_estimated_minutes: int
    overall_accuracy_pct: float
    most_time_consuming_project: Optional[str] = None
    most_accurate_estimates: Optional[str] = None


# --- Project Progress / Burndown ---


class BurndownEntry(BaseModel):
    date: date
    # `remaining` and `completed_cumulative` are floats since the
    # PrismTask-timeline-class scope (P9 option a) — a task with
    # progress_percent = 60 contributes 0.6 of a unit. `added` stays
    # int because it counts task rows created on a day, not fractional
    # progress.
    remaining: float
    completed_cumulative: float
    added: int


class ProjectProgressResponse(BaseModel):
    project_name: str
    total_tasks: int
    # `completed_tasks` is the project-wide sum of fractional
    # contributions (`progress_percent / 100` per task, with a fallback
    # of 1.0 for binary-DONE rows). Float because partial-progress rows
    # land non-integer values; legacy int-only consumers should round
    # at the call site.
    completed_tasks: float
    burndown: list[BurndownEntry]
    velocity: float
    projected_completion: Optional[date] = None
    is_on_track: bool


# --- Habit Correlations ---


class HabitCorrelation(BaseModel):
    habit: str
    done_productivity: float
    not_done_productivity: float
    correlation: str  # "positive", "negative", "neutral"
    interpretation: str


class HabitCorrelationResponse(BaseModel):
    correlations: list[HabitCorrelation]
    top_insight: str
    recommendation: str


# --- Summary ---


class TodaySummary(BaseModel):
    completed: int
    remaining: int
    score: float


class WeekSummary(BaseModel):
    completed: int
    remaining: int
    score: float
    trend: str


class MonthSummary(BaseModel):
    completed: int
    remaining: int
    score: float


class StreakSummary(BaseModel):
    current_productive_days: int
    longest_productive_days: int


class HabitSummary(BaseModel):
    completion_rate_7d: float
    completion_rate_30d: float


class AnalyticsSummaryResponse(BaseModel):
    today: TodaySummary
    this_week: WeekSummary
    this_month: MonthSummary
    streaks: StreakSummary
    habits: HabitSummary
